source/grid/grid_in.cc

```diff
@@ -3252,6 +3252,88 @@ GridIn<dim, spacedim>::read_partitioned_msh(const std::string &file_prefix,
           coords[3 * i + d];
     }
 
+  // --- Collect physical groups for boundary and manifold info ---
+  std::map<std::pair<int, int>, types::boundary_id> entity_to_boundary;
+  std::map<std::pair<int, int>, types::material_id> entity_to_material;
+
+
+  std::vector<std::pair<int, int>> physical_groups;
+  gmsh::model::getPhysicalGroups(physical_groups);
+
+  for (const auto &pg : physical_groups)
+    {
+      int physical_dim = pg.first;
+      int physical_tag = pg.second;
+
+      std::vector<int> physical_entities;
+      gmsh::model::getEntitiesForPhysicalGroup(physical_dim,
+                                               physical_tag,
+                                               physical_entities);
+
+      for (int ent : physical_entities)
+        {
+          // Assign boundary ID for faces (dim-1)
+          if (physical_dim == dim - 1)
+            entity_to_boundary[{physical_dim, ent}] = physical_tag;
+          if (physical_dim == dim) // Volume materials
+            entity_to_material[{physical_dim, ent}] = physical_tag;
+        }
+    }
+
+  // --- Build boundary face map for all (dim-1) entities ---
+
+  std::map<std::set<unsigned int>, types::boundary_id> boundary_face_map;
+
+  // Process all (dim-1)-dimensional entities for boundary/manifold assignment
+  for (const auto &e : entities)
+    {
+      if (e.first == dim - 1) // Faces in 3D, edges in 2D, vertices in 1D
+        {
+          types::boundary_id boundary_id = numbers::invalid_boundary_id;
+
+          // Check for boundary ID
+          if (auto it = entity_to_boundary.find({e.first, e.second});
+              it != entity_to_boundary.end())
+            boundary_id = it->second;
+
+          if (boundary_id != numbers::invalid_boundary_id)
```

Comment:
Since the condition of the last `if` can only be true if the one above evaluated to true, shouldn't we try to merge these statements into a simpler one:
```suggestion
      if (auto it = entity_to_boundary.find({e.first, e.second});
          e.first == dim - 1 && it != entity_to_boundary.end()) // Faces in 3D, edges in 2D, vertices in 1D
        {
          const types::boundary_id boundary_id = it->second;
```